lib/ui/painting/image_encoding.cc

```diff
@@ -96,6 +81,64 @@ sk_sp<SkImage> ConvertToRasterImageIfNecessary(sk_sp<SkImage> image,
   return snapshot->makeRasterImage();
 }
 
+void ConvertImageToRaster(sk_sp<SkImage> image,
+                          std::function<void(sk_sp<SkImage>)> encode_task,
+                          fml::RefPtr<fml::TaskRunner> gpu_task_runner,
+                          fml::RefPtr<fml::TaskRunner> io_task_runner,
+                          GrContext* resource_context,
+                          fml::WeakPtr<SnapshotDelegate> snapshot_delegate) {
+  // Check validity of the image.
+  if (image == nullptr) {
+    FML_LOG(ERROR) << "Image was null.";
+    encode_task(nullptr);
+    return;
+  }
+
+  auto dimensions = image->dimensions();
+
+  if (dimensions.isEmpty()) {
+    FML_LOG(ERROR) << "Image dimensions were empty.";
+    encode_task(nullptr);
+    return;
+  }
+
+  SkPixmap pixmap;
+  if (image->peekPixels(&pixmap)) {
+    // This is already a raster image.
+    encode_task(image);
+    return;
+  }
+
+  if (sk_sp<SkImage> raster_image = image->makeRasterImage()) {
+    // The image can be converted to a raster image.
+    encode_task(raster_image);
+    return;
+  }
+
+  // Cross-context images do not support makeRasterImage. Convert these images
+  // by drawing them into a surface.  This must be done on the GPU thread
+  // to prevent concurrent usage of the image on both the IO and GPU threads.
+  gpu_task_runner->PostTask([image, encode_task = std::move(encode_task),
+                             resource_context, snapshot_delegate,
+                             io_task_runner]() {
+    sk_sp<SkImage> raster_image =
+        snapshot_delegate->ConvertToRasterImage(image);
+
+    io_task_runner->PostTask([image, encode_task = std::move(encode_task),
+                              raster_image = std::move(raster_image),
+                              resource_context]() mutable {
+      if (!raster_image) {
```

Comment:
Can this check not be performed in the `gpu_task_runner` closure with the hop to the IO task runner elided in case of successful raster image acquisition? Just a minor optimization. This works fine too.